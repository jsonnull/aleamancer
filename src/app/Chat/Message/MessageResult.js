/* @flow */
import React from 'react'
import type { MessageResult } from 'types'
import styles from './style.css'

type RollsProps = {
  rolls: Array<number>
}
const Rolls = (props: RollsProps) => {
  return <span>
    <i className='fa fa-cube'></i>
    {' '}
    {props.rolls.map((roll, i, arr) => (
      <span key={i}>
        <span className={styles.roll}>
          { roll }
        </span>
        {(i + 1 !== arr.length) ? ' + ' : null}
      </span>
    ))}
  </span>
}

type ModProps = {
  mod: number
}
const Mod = (props: ModProps) => {
  if (props.mod == 0) {
    return null
  }

  return <span>
    {' + '}
    <span className={styles.mod}>{props.mod}</span>
  </span>
}

type Props = {
  result: ?MessageResult
}
const Result = (props: Props) => {
  if (!props.result) {
    return null
  }

  let { mod, rolls, total } = props.result

  return (
    <div className={styles.result}>
      <Rolls rolls={rolls} />
      <Mod mod={mod} />
      {' = '}
      <span className={styles.total}>{total}</span>
    </div>
  )
}

export default Result
