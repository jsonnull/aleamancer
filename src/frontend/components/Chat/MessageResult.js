// @flow
import React from 'react'
import styled from 'styled-components'
import type { MessageResult, Roll } from 'common/types'

const RollSpan = styled.span`
  font-weight: bold;
`

type RollProps = {
  result: number,
  operation: '+' | '-',
  color?: 'inherit' | 'green' | 'red',
  showOperation?: boolean
}
const Result = (props: RollProps) => {
  const { result, operation, color = 'inherit', showOperation = true } = props

  const shownOperation = showOperation ? ` ${operation} ` : ''

  return (
    <span>
      {shownOperation}
      <RollSpan style={{ color }}>{result}</RollSpan>
    </span>
  )
}

type RollContainerProps = {
  roll: Roll | number,
  isFirst: boolean
}
const RollContainer = (props: RollContainerProps) => {
  const { roll, isFirst } = props

  // Handle the general case where the roll is a constant
  if (typeof roll === 'number') {
    return <Result result={roll} operation="+" showOperation={!isFirst} />
  }

  const { die, result, operation } = roll

  let color = 'inherit'
  if (result === die) {
    color = 'green'
  } else if (result === 1) {
    color = 'red'
  }

  return (
    <Result
      result={result}
      color={color}
      operation={operation}
      showOperation={!isFirst}
    />
  )
}

const ResultsContainer = styled.div`
  padding: 0 1rem 0.2rem;
`
const Total = styled.span`
  font-weight: bold;
`

type Props = {
  result: ?MessageResult
}
const Results = (props: Props) => {
  if (!props.result) {
    return null
  }

  const results = props.result
  const total = results.reduce((acc, roll) => {
    if (typeof roll === 'number') {
      return acc + roll
    } else {
      return acc + roll.result
    }
  }, 0)

  return (
    <ResultsContainer>
      <i className="fa fa-cube" />{' '}
      {results.map((result, i) => (
        <span key={i}>
          <RollContainer roll={result} isFirst={i === 0} />
        </span>
      ))}
      {' = '}
      <Total>{total}</Total>
    </ResultsContainer>
  )
}

export default Results
