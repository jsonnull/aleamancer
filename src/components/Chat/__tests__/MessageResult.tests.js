// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import MessageResult from '../MessageResult.js'

describe('Chat MessageResult component', () => {
  it('renders correctly', () => {
    const result = [
      {
        die: 6,
        result: 4,
        operation: '+'
      }
    ]
    const tree = renderer.create(<MessageResult result={result} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
