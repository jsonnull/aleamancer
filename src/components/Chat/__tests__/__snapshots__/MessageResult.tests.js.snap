// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat MessageResult component renders correctly 1`] = `
<div
  className="sc-bwzfXH dslxFQ"
>
  <i
    className="fa fa-cube"
  />
   
  <span>
    <span>
      
      <span
        className="sc-bdVaJa hCboOS"
        style={
          Object {
            "color": "inherit",
          }
        }
      >
        4
      </span>
    </span>
  </span>
   = 
  <span
    className="sc-htpNat joRjxB"
  >
    4
  </span>
</div>
`;
